gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/DataFlowTopology.java

```diff
@@ -0,0 +1,150 @@
+    for (ReplicationReplica replica : builder.replicas) {
+      Preconditions.checkArgument(routesConfig.hasPath(replica.getReplicationName()),
+          "can not find route for replica " + replica.getReplicationName());
+
+      List<ReplicationData<CopyableDataset>> copyFromReplica = new ArrayList<ReplicationData<CopyableDataset>>();
+
+      List<String> copyFromStrings = routesConfig.getStringList(replica.getReplicationName());
+
+      for (String copyFromName : copyFromStrings) {
+        Preconditions.checkArgument(!copyFromName.equals(replica.getReplicationName()),
+            "can not have same name in both destination and copy from list " + copyFromName);
+        // copy from source
+        if (copyFromName.equals(ReplicationUtils.REPLICATION_SOURCE)) {
+          copyFromReplica.add(builder.source);
+        }
+        // copy from other replicas
+        else if(replicasMap.containsKey(copyFromName)){
+          copyFromReplica.add(replicasMap.get(copyFromName));
+        }
+        else{
+          throw new IllegalArgumentException("can not find replica with name " + copyFromName);
+        }
+      }
+      
+      CopyRouteBuilder routeBuilder = new CopyRouteBuilder();
+      routeBuilder.withCopyDestination(replica);
+      routeBuilder.withCopyFrom(copyFromReplica);
+      this.routes.add(routeBuilder.build());
+    }
+  }
+
+  public static class CopyRoute {
+    @Getter
+    private final ReplicationReplica copyDestination;
+
+    @Getter
+    private final List<ReplicationData<CopyableDataset>> copyFrom;
+    
+    private CopyRoute(CopyRouteBuilder builder){
+      this.copyDestination = builder.copyDestination;
+      this.copyFrom = builder.copyFrom;
+    }
+
+    @Override
+    public String toString() {
+      StringBuilder sb = new StringBuilder();
+      sb.append(copyDestination.getReplicationName() + ":");
+      for (ReplicationData<CopyableDataset> r : copyFrom) {
+        sb.append(r.getReplicationName() + ",");
+      }
+
+      return sb.toString();
+    }
+  }
+  
+  public static class CopyRouteBuilder{
+    private ReplicationReplica copyDestination;
+    
+    private List<ReplicationData<CopyableDataset>> copyFrom;
+    
+    public CopyRouteBuilder withCopyDestination(ReplicationReplica copyDest) {
+      this.copyDestination = copyDest;
+      return this;
+    }
+
+    public CopyRouteBuilder withCopyFrom(List<ReplicationData<CopyableDataset>> copyFrom) {
+      this.copyFrom = copyFrom;
+      return this;
+    }
+
+    public CopyRoute build() {
+      return new CopyRoute(this);
+    }
+  }
+
+  public static class Builder {
+
+    private ReplicationSource source;
+
+    private List<ReplicationReplica> replicas;
+
+    private Config topologyConfig;
+
+    public Builder withReplicationSource(ReplicationSource source) {
+      this.source = source;
+      return this;
+    }
+
+    public Builder withReplicationReplicas(List<ReplicationReplica> replicas) {
```

Comment:
When dealing with lists in builders, I think it is useful to have a method that can add individual elements to the list rather then require populating the list separately. For example, the API can be "addReplica(Replica replica)".